File: cpp/med/Question1052.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int maxSatisfied(vector<int> &customers, vector<int> &grumpy, int minutes) {
        int max_s = 0, sum = 0;
        for (int right = 0; right < customers.size(); right++) {
            if (grumpy[right] == 1) {
                sum += customers[right];
            }
            if (right < minutes - 1) {
                continue;
            }
            max_s = max(max_s, sum);
            if (grumpy[right - minutes + 1] == 1) {
                sum -= customers[right - minutes + 1];
            }
        }
        int ans = 0;
        for (int i = 0; i < customers.size(); i++) {
            if (grumpy[i] == 0) {
                ans += customers[i];
            }
        }
        return ans + max_s;
    }
};

File: python/med/Question1052.py
from typing import List


class Solution:
    """
    方法一：滑动窗口
        计算能挽回的顾客的最大值
    """

    def maxSatisfied(
        self, customers: List[int], grumpy: List[int], minutes: int
    ) -> int:
        n = len(customers)
        ans = 0
        save = 0
        other = 0
        for i in range(n):
            if grumpy[i] == 1:
                save += customers[i]
            else:
                other += customers[i]
            if i - minutes + 1 < 0:
                continue
            else:
                ans = max(ans, save)
                if grumpy[i - minutes + 1] == 1:
                    save -= customers[i - minutes + 1]
        return other + ans


print(Solution().maxSatisfied([1, 0, 1, 2, 1, 1, 7, 5], [0, 1, 0, 1, 0, 1, 0, 1], 3))
# [1, 0, 1, 2, 1, 1, 7, 5]
# [0, 1, 0, 1, 0, 1, 0, 1]
